src/ws/Socket.ts

```diff
@@ -431,8 +431,12 @@ export class Socket extends EventEmitter {
             this.call(authMethod, this._authpacket, { force: true })
             .then(result => this.emit('authresult', result))
             .then(bang)
-            .catch(() => {
-                this.emit('error', new AuthenticationFailedError('?'));
+            .catch((e: Error) => {
+                let message = 'Authentication Failed, please check your credentials.'
```

Comment:
strange, I could swear I submitted that yesterday... GitHub, please.

Reply:
I think this is a huge improvement from the previous state where it just errored with `?`. We can add additional messages and helpful hints here in the future.

Reply:
I would much prefer either you include all error types the server can return so your code serves as documentation, of those I've seen at least `UNOTFOUND` for invalid auth key, `UACCESS` for not authorized to view channel chat in test stream mode and `ETIMEOUT` (probably async database call failing) do exist, and that you emit an error of the correct type, or that you do `this.emit('error', new AuthenticationFailedError(e.message));` instead, rather than getting `Authentication Failed: User not found. Please check our guide at: https://aka.ms/unotfound` for the `UNOTFOUND` response and `Authentication Failed, please check your credentials.` for all unhandled cases.

Reply:
Error classes are easy to add as an addition here.

Reply:
except that they don't quite work at runtime... at least not in JavaScript https://www.typescriptlang.org/docs/handbook/advanced-types.html#user-defined-type-guards

Reply:
@ProbablePrime we could probably just add a `code` property or something to the error message or we subclass the existing error.